h2o-core/src/main/java/water/H2ONode.java

```diff
@@ -471,20 +509,28 @@ public void sendMessage(ByteBuffer bb, byte msg_priority) {
       if( msg_priority > bb.limit() ) msg_priority = (byte)bb.limit();
       bb.position(msg_priority);
 
-      _msgQ.put(bb);
+      _outgoingMsgQ.put(bb);
     }
 
-    private final PriorityBlockingQueue<ByteBuffer> _msgQ
-      = new PriorityBlockingQueue<>(11,new Comparator<ByteBuffer>() {
-          // Secret back-channel priority: the position field (capped at bb.limit)
-          @Override public int compare( ByteBuffer bb1, ByteBuffer bb2 ) { return bb1.position() - bb2.position(); }
-        });
+    private boolean isActive() {
+      return _sendThread == this || (_sendThread == null && isClientNode());
+    }
+
+    // We deliver messages to regular nodes only if the are part of the cloud
+    // and to always to clients 
+    private boolean keepSending() {
+      return !isRemovedFromCloud() || isClientNode();
+    }
 
     @Override public void run(){
       try {
-        while (!_stopRequested) {            // Forever loop
+        while (isActive()) {            // Forever loop
           try {
-            ByteBuffer bb = _msgQ.take(); // take never returns null but blocks instead
+            ByteBuffer bb = _outgoingMsgQ.take(); // take never returns null but blocks instead
+            if (! isActive()) {
```

Comment:
Shouldn't we test this on line 538, before we actually try to send the data?

Reply:
The design is as follows: if we are active at the time of consuming the message, we will proceed to the sending stage. Then we will try to deliver the message based on the decision of the `keepSending` function which doesn't care anymore if the thread is active or not, it is only driven by whether the message should be (still) delivered or not.